lib/corefoundation/preferences.rb

```diff
@@ -23,49 +20,133 @@ module CF
   attach_function "CFPreferencesSetValue", %i{key value application_id username hostname}, :void
   attach_function "CFPreferencesAppSynchronize", [:application_id], :bool
 
+  # Interface to the preference utilities from Corefoundation.framework.
+  # Documentation at https://developer.apple.com/documentation/corefoundation/preferences_utilities
+  #
   class Preferences
     CURRENT_USER = CF.kCFPreferencesCurrentUser
     ALL_USERS = CF.kCFPreferencesAnyUser
     CURRENT_HOST = CF.kCFPreferencesCurrentHost
     ALL_HOSTS = CF.kCFPreferencesAnyHost
 
-    class << self
-      def get(key, application_id, username = nil, hostname = nil)
-        # TODO: Check usecase for `CFPreferencesCopyAppValue`
+    # Returns the output from `CFPreferencesCopyValue` call after converting it to ruby type.
+    #
+    # @param [String] key Preference key to read
+    # @param [String] application_id Preference domain for the key
+    # @param [String, Symbol] username Domain user (current, any or a specific user)
+    # @param [String, Symbol] hostname Hostname (current, all hosts or a specific host)
+    #
+    # @return [VALUE] Preference value returned from the `CFPreferencesCopyValue` call.
+    #
+    def self.get(key, application_id, username = nil, hostname = nil)
+      username ||= CURRENT_USER
+      hostname ||= ALL_HOSTS
```

Comment:
This library is supposed to be a vanilla/unopinionated/direct mapping to CF APIs, so let's avoid setting defaults here and set them where this is consumed instead, i.e., `macos_userdefaults` resource/wrapper

Reply:
From what I could figure, this is a default for defaults preferences by corefoundation when accessed via the high-level counterparts of copy and write APIs. Passing nil should also result in the same operation, but it becomes implicit or ambiguous if you want to call it that. "Setting" as default for our arguments is more a way of expressing that these are the defaults rather than trying to enforce anything.

Reply:
The reason why I mention it being "appropriate" in this PR's thread here: https://github.com/chef/corefoundation/pull/11#discussion_r682321134

Reply:
Because of a wider possible use of this library and possible change of defaults in the future, let's avoid setting defaults. This is also related to the behaviour of `CFPreferencesCopyAppValue` and historic `defaults` utility usage in `macos_userdefaults` resource.

Reply:
`CFPreferencesCopyAppValue` is most likely to not change very soon, according to a comment inside the corefoundation source code. Which also makes sense. Please refer to the list in https://github.com/chef/corefoundation/pull/11#discussion_r684119823 which has the right order. 

And to reiterate, the ones "set" are not setting anything or changing the behavior, that's an idiomatic way of expressing that those are the defaults.

Reply:
> `CFPreferencesCopyAppValue` is most likely to not change very soon, according to a comment inside the corefoundation source code.
Could you point to that?

Also, we have to consider wider/generic possible usage of this library. Having defaults like this will defeat that purpose.

Reply:
> Could you point to that?

I can't link to the comment in a raw file. But let me find it and send over a screenshot. Also, that should be `CFPreferencesCopyValue` and not `CFPreferencesCopyAppValue`. The latter utilises a search list with a particular order. That list is in a comment in this PR.

> Also, we have to consider wider/generic possible usage of this library. Having defaults like this will defeat that purpose.

From the library perspective, we have `CF.CFPreferencesCopyValue` and `CF.CFPreferencesCopyAppValue` available for use. So if you ask me, have any of the get or set methods would not make sense unless there's a specific use case we are trying to cover. what do you think?

Reply:
To add on to that, I remember saying this a few days ago that we need to have a discussion around how we want this library to grow eventually. Do we want to keep adding utilities for the use cases that we encounter along the way or keep it minimal and only act as an interface to the corefoundation framework.

Reply:
> From the library perspective, we have CF.CFPreferencesCopyValue and CF.CFPreferencesCopyAppValue available for use. So if you ask me, have any of the get or set methods would not make sense unless there's a specific use case we are trying to cover. what do you think?

The use case is that we want to add a thin ruby layer to interact with CF APIs. That being said, it's not required to bring all the baggage of C APIs into this layer, hence, we are developing generic utility methods that expose enough options in an idiomatic way without being opinionated for it's usage in `macos_userdefaults` resource.

> To add on to that, I remember saying this a few days ago that we need to have a discussion around how we want this library to grow eventually

I believe from the time we were deciding b/w objC wrapper or ruby-ffi approach, it's clear that this is meant to be generic, thin and basis for further integration with other OSX APIs. We can stick to this philosophy/design.